### ISSUE TITLE
[Clang] AST has inconsistent ImplicitCast expression type for `ParenType` based on qualifiers

### ISSUE THREAD

**Issue by Jezurko** [llvm/llvm-project#98637]
### Description
I have discovered an inconsistency in `ImplicitCastExpr` that happens when `ParenType` is involved.
Consider the following program:
```
int main() {
   int (a) = 0;
   a + 0;
}
```
Now when you dump the AST for the binary expression, you get:
```
    `-BinaryOperator 0x55790e9b6fb8 <line:3:5, col:9> 'int' '+'
      |-ImplicitCastExpr 0x55790e9b6fa0 <col:5> 'int' <LValueToRValue>
      | `-DeclRefExpr 0x55790e9b6f60 <col:5> 'int' lvalue Var 0x55790e9b6ec0 'a' 'int'
      `-IntegerLiteral 0x55790e9b6f80 <col:9> 'int' 0

```
Which seems fine. But actually dumping the result type of the `ImplicitCast` we get this:
```
ParenType 0x55556b68fd10 'int' sugar
`-BuiltinType 0x55556b
```

Now if we modify the program by adding `const` to the integer variable:
```
int main() {
   const int (a) = 0;
   a + 0;
}
```
The AST for the binary expression looks almost the same:
```
`-BinaryOperator 0x557c6784ffb8 <line:3:5, col:9> 'int' '+'
      |-ImplicitCastExpr 0x557c6784ffa0 <col:5> 'int' <LValueToRValue>
      | `-DeclRefExpr 0x557c6784ff60 <col:5> 'const int' lvalue Var 0x557c6784fec0 'a' 'const int'
      `-IntegerLiteral 0x557c6784ff80 <col:9> 'int' 0
```
But dumping the result type of the `Implicit cast yields an unexpected result (at least in my opinion):
```
BuiltinType 0x55556b62e0e0 'int'
```
### Cause
When I looked into why this happens, I discovered that in `clang/lib/Sema/SemaExpr.cpp` in the `ExprResult Sema::DefaultLvalueConversion(Expr *E)` function the qualifiers get stripped. But stripping the qualifiers also removes the `ParenType` from the type:
![lldb screenshot showing that stripping qualifiers removes the `ParenType`](https://github.com/user-attachments/assets/060b1b92-a533-4fd3-a72c-2e1f16df151d)
### Expected behaviour
I would expect these to behave the same in relation to the `ParenType`, so that either both cases keep it, or both remove it.
Personally, I would prefer the solution that it gets stripped by the cast, as it would make my life easier, but I am not sure if that's the correct way.

**Comment by shafik** [comment]
CC @AaronBallman

**Comment by Jezurko** [comment]
We are building [VAST](https://github.com/trailofbits/vast) - an MLIR based compiler for verification uses.
In our CodeGen, we consume the AST and generate our HighLevel IR from it. We obtain this SSA-like IR by translating the AST, including types, basically 1:1. The MLIR Operations have verifiers that check if the operation is valid (which is usually checking if types are correct). We have discovered that some of our verifiers were failing and that ParenType was involved - because the expected type was wrapped in ParenType and `mlir::isa` (which is internally used by the verifiers generated from TableGen) for the expected type fails on it.
I can update the verifiers on our side to work with ParenType, but this inconsistent behaviour seemed like a bug to me.

**Comment by AaronBallman** [comment]
Getting the unqualified type strips sugar, and parens are considered sugar, which is why that gets stripped: https://github.com/llvm/llvm-project/blob/ef51e617c43ef4248973f07e0a711f69d85c948c/clang/lib/AST/Type.cpp#L520 The reason the qualifier matters is because we take a different code path if there are local qualifiers: https://github.com/llvm/llvm-project/blob/ef51e617c43ef4248973f07e0a711f69d85c948c/clang/include/clang/AST/Type.h#L7839

The documentation for `getUnqualifiedType()` is clear that it strips sugar until it finds the least-sugared unqualified type. `getSplitUnqualifiedType()` gets the unqualified type by removing as little sugar as possible -- but that looks like it will also strip the `ParenType` in this case.

One thing to try would be to work on the canonical type within `DefaultLValueConversion` when calling `QualType T = E->getType();`. I see we have code in that function like `T == Context.OverloadTy` which won't work properly because it's not looking at the canonical type, but the `T->isRecordType()` and similar calls are looking at the canonical type.

**Comment by mizvekov** [comment]
Yeah, it seems your verifier may only care about canonical types, so it's just easier to work on the canonical type, and by doing that you will drop a lot of complexity on the floor.

Otherwise, the way we deal in clang with these type adjustments has some pretty ingrained limitations.

In some cases, we try to preserve the original type when we do these adjustments. For example, we have an AdjustedType, which represents these type adjustments by holding the original type in addition to the adjusted type.

But if we started using these everywhere we simply need an unqualified type, this would have a performance hit and a usability hit.

**Comment by Jezurko** [comment]
Thank you for the responses!

> Yeah, it seems your verifier may only care about canonical types, so it's just easier to work on the canonical type, and by doing that you will drop a lot of complexity on the floor.

Yes, I can do that. It makes stuff a bit more complicated for me, but that wasn't my main point in this issue - I raised it because of the strangely inconsistent behaviour.

> But if we started using these everywhere we simply need an unqualified type, this would have a performance hit and a usability hit.

Agreed! I also don't think it makes sense to keep `QualType`/sugar when it's not needed. That's partially why I raised this issue. The current behaviour of the `LValueToRValue` cast is, that if the type that is being casted does not have qualifiers (based on the `QualType` method `hasQualifiers`), the sugar is kept. But since for the qualified case the sugar is clearly removed, I would assume it is safe to do so even in the unqualified case.

>  I see we have code in that function like T == Context.OverloadTy which won't work properly because it's not looking at the canonical type, but the T->isRecordType() and similar calls are looking at the canonical type.

Ah, I didn't notice that before, because I was looking on 18.1 release where the comparison was still made with `E->getType()` (the rest of the condition already uses `T`). Seems to me like in the past someone was maybe already anticipating that `T` will not have the qualifiers?
